### PR TITLE
Return metrics on custom object evaluation

### PR DESCRIPTION
**Pull request by rola93**
- Return a list of dict containing the metrics that are currently being displayed on terminal. 
- Refactor code to avoid unnecessary duplicates. 
- Show more information on unexpected flows

### CODE REVIEW

**Comment by auphofBSF** [comment]
Very useful, Thank You @rola93 
I have tested this locally and confirm the returned metrics dictionary is as displayed by 
`trainer.evaluateModel(...)` 
Your work on this was timely as I was just looking to use this to further automate workflow in Object Detection training, specifically resuming training. https://github.com/OlafenwaMoses/ImageAI/issues/298#issuecomment-522796921

File: imageai/Detection/Custom/__init__.py [hunk 1]
```diff
@@ -423,44 +426,31 @@ def evaluateModel(self, model_path, json_path, batch_size=4, iou_threshold=0.5,
                     average_precisions = evaluate(infer_model, valid_generator, iou_threshold=iou_threshold,
                                                   obj_thresh=object_threshold, nms_thresh=nms_threshold)
 
+                    result_dict = {
+                        'model_file': model_path,
```
Comment (auphofBSF) [inline imageai/Detection/Custom/__init__.py]: model_file' : os.path.join(model_path, model_file),

File: imageai/Detection/Custom/__init__.py [hunk 2]
```diff
@@ -423,44 +426,31 @@ def evaluateModel(self, model_path, json_path, batch_size=4, iou_threshold=0.5,
...
+                        'using_iou': iou_threshold,
+                        'using_object_threshold': object_threshold,
+                        'using_non_maximum_suppression': nms_threshold,
+                        'average_precision': dict()
+                    }
                     # print the score
                     print("Model File: ", model_path, '\n')
```
Comment (auphofBSF) [inline imageai/Detection/Custom/__init__.py]: print("Model File: ", os.path.join(model_path, model_file), '\n')

File: imageai/Detection/Custom/__init__.py (hunk 1 above)
Comment (auphofBSF) [inline imageai/Detection/Custom/__init__.py]: model_file' : os.path.join(model_path, model_file),

**Comment by auphofBSF** [comment]
> This is a brilliant work. Thanks for the effort on this. It will be better if you can effect the following changes:
> 
> * add `print` of  **mAP** evaluation result for each model if a directory path is parsed

### Suggestions for updating documentation,

- @rola93 do you want to incorporate in your PR #302 
- @OlafenwaMoses is the only documentation location ?

https://github.com/OlafenwaMoses/ImageAI/blob/master/imageai/Detection/Custom/CUSTOMDETECTIONTRAINING.md#evaluating-your-saved-detection-models-map
```
from imageai.Detection.Custom import DetectionModelTrainer

trainer = DetectionModelTrainer()
trainer.setModelTypeAsYOLOv3()
trainer.setDataDirectory(data_directory="/content/hololens")
evaluate_metrics = trainer.evaluateModel(model_path="/content/hololens/models", 
                               json_path="/content/hololens/json/detection_config.json", 
                               iou_threshold=0.5, 
                               object_threshold=0.3, 
                               nms_threshold=0.5)
```

### Console stdout and stderr

```
Using TensorFlow backend.
Starting Model evaluation....
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/imageai-2.1.3-py3.6.egg/imageai/Detection/Custom/yolo.py:24: to_float (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.cast instead.
/usr/local/lib/python3.6/dist-packages/keras/engine/saving.py:292: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.
  warnings.warn('No training configuration found in save file: '
Model File:  /content/hololens/models/detection_model-ex-02--loss-7.47.h5 

Using IoU :  0.5
Using Object Threshold :  0.3
Using Non-Maximum Suppression :  0.5
hololens: 0.8252
mAP: 0.8252
===============================
Model File:  /content/hololens/models/detection_model-ex-01--loss-25.31.h5 

Using IoU :  0.5
Using Object Threshold :  0.3
Using Non-Maximum Suppression :  0.5
hololens: 0.4390
mAP: 0.4390
===============================
Model File:  /content/hololens/models/detection_model-ex-03--loss-6.20.h5 

Using IoU :  0.5
Using Object Threshold :  0.3
Using Non-Maximum Suppression :  0.5
hololens: 0.8415
mAP: 0.8415
===============================
```
### Dictionary view of metrics
`evaluate_metrics`

```
[{'average_precision': {'hololens': 0.8252334437735249},
  'map': 0.8252334437735249,
  'model_file': '/content/hololens/models/detection_model-ex-02--loss-7.47.h5',
  'using_iou': 0.5,
  'using_non_maximum_suppression': 0.5,
  'using_object_threshold': 0.3},
 {'average_precision': {'hololens': 0.4389696930319986},
  'map': 0.4389696930319986,
  'model_file': '/content/hololens/models/detection_model-ex-01--loss-25.31.h5',
  'using_iou': 0.5,
  'using_non_maximum_suppression': 0.5,
  'using_object_threshold': 0.3},
 {'average_precision': {'hololens': 0.8414673569376486},
  'map': 0.8414673569376486,
  'model_file': '/content/hololens/models/detection_model-ex-03--loss-6.20.h5',
  'using_iou': 0.5,
  'using_non_maximum_suppression': 0.5,
  'using_object_threshold': 0.3}]
```
### Example of use
```
for metric in evaluate_metrics:
  print("Model:{modelFile} \t Average mAP:{mAP:.4f}".format(modelFile=metric['model_file'],mAP=metric['map']))

print('--------------------------------------------------------------------')
for metric in evaluate_metrics:
  print("Model:{modelFile}".format(modelFile=metric['model_file']))
  for theclass in metric['average_precision']:
      print("\t\t\t\t\t\t\t\t Item:{item} \t mAP:{mAP:.4f}".format(item=theclass, mAP=metric['average_precision'][theclass]))
```

```
Model:/content/hololens/models/detection_model-ex-02--loss-7.47.h5 	 Average mAP:0.8252
Model:/content/hololens/models/detection_model-ex-01--loss-25.31.h5 	 Average mAP:0.4390
Model:/content/hololens/models/detection_model-ex-03--loss-6.20.h5 	 Average mAP:0.8415
--------------------------------------------------------------------
Model:/content/hololens/models/detection_model-ex-02--loss-7.47.h5
								 Item:hololens 	 mAP:0.8252
Model:/content/hololens/models/detection_model-ex-01--loss-25.31.h5
								 Item:hololens 	 mAP:0.4390
Model:/content/hololens/models/detection_model-ex-03--loss-6.20.h5
								 Item:hololens 	 mAP:0.8415
```

File: imageai/Detection/Custom/__init__.py (hunk 2 above)
Comment (rola93) [reply imageai/Detection/Custom/__init__.py]: my apologies for this, I think the error is not only here but also some previous lines, I'm working on that rigth now

**Comment by rola93** [comment]
@auphofBSF, I'm sorry but I didn't get your point with respect to updating the docs

I was wondering if shouldn't add a `verbose` parameter to avoid printing always, what do you think?

**Comment by auphofBSF** [comment]
> @auphofBSF, I'm sorry but I didn't get your point with respect to updating the docs
> 
> I was wondering if shouldn't add a `verbose` parameter to avoid printing always, what do you think?

I like your thinking but would like to suggest the following to improve overall monitoring
(@OlafenwaMoses I would appreciate your thoughts on this)

1. **A more consistent  and multi-functional global level of verbosity**,  `print(...)`  in project be converted to using the python logging. module 
```
import logging
logging.info(......)
logging.debug(.....)
```
In that way verbosity  can be globally configured for each application usage of imageai and granularly adjusted for each class with a diverse set of logging options up to the final user

2. **More detailed progress metrics**, I was looking through all the code for custom object detection and looking for some more granular level of metrics on training or evaluation progress.  I extensively use the module [prometheus_client](https://github.com/prometheus/client_python ) `import 
 prometheus_client`   for monitoring of long running python applications, its lightweight and easy to add and just works.

I look forward to other suggestion and possibly putting this on feature_request, or enhancements list and to other suggestions and willing to help put some of this in place

**Comment by rola93** [comment]
I agree with you, print are not the best option as I see it, and maybe moving to logging isa better option, but it goes out of scope for this threaad, maybe we can open an issue to keep discussing it. 

with respect to the more detailed logging on training, I've never heard about prometheus client, I will read to see what is it about, but at first sight, monitoring training looks like very similar to what Tensorboard callback does. It was added on [291](https://github.com/OlafenwaMoses/ImageAI/pull/291) but only for custom object detection training (maybe should add it to the image classifier training too to keep it uniform)

Once again, this second point is a nice discussion which is out of scope for this thread. Lets open an issue to discuss it :D

**Comment by auphofBSF** [comment]
> @auphofBSF, I'm sorry but I didn't get your point with respect to updating the docs
> 
Hello @rola93 with regard to the [documentation](https://github.com/OlafenwaMoses/ImageAI/pull/302#issuecomment-523227035) I was responding to @OlafenwaMoses  comment 
https://github.com/OlafenwaMoses/ImageAI/pull/302#pullrequestreview-277393518 
> add print of mAP evaluation result for each model if a directory path is parsed

As I was finishing my response I thought well this or something like this really needs to go into the documentation https://github.com/OlafenwaMoses/ImageAI/blob/master/imageai/Detection/Custom/CUSTOMDETECTIONTRAINING.md#evaluating-your-saved-detection-models-map.

I thought if you pasted in or made your own variation 
> add print of mAP evaluation result for each model if a directory path is parsed

into **/imageai/Detection/Custom/CUSTOMDETECTIONTRAINING.m**  then it could all be incorporated within the same PR #302
Willing to assist however

**Comment by auphofBSF** [comment]
> With respect to the more detailed logging on training, I've never heard about prometheus client, I will read to see what is it about, but at first sight, monitoring training looks like very similar to what Tensorboard callback does. It was added on [291](https://github.com/OlafenwaMoses/ImageAI/pull/291) but only for custom object detection training (maybe should add it to the image classifier training too to keep it uniform)
> 
Agreed these are both out of scope, I will create  Issue/Enhancements for each of them.   
The Tensorboard support you added in #291 certainly looks great. It offers the prospect of  real in-depth and very useful training monitoring. I am a relative Newbie to DL and all the tools, so look forward to looking at that, will continue discussions in the new to be created issue/enhancement.

**Comment by rola93** [comment]
> Hello @rola93 with regard to the [documentation](https://github.com/OlafenwaMoses/ImageAI/pull/302#issuecomment-523227035) I was responding to @OlafenwaMoses comment
> [#302 (review)](https://github.com/OlafenwaMoses/ImageAI/pull/302#pullrequestreview-277393518)

ok, I think I got it now, thanks for the clarification @auphofBSF 


>     * add `print` of  **mAP** evaluation result for each model if a directory path is parsed

@OlafenwaMoses I think it's already there. Consider that the logic was modified a little bit to avoid duplicate code, so now if a filename is given, it is put in a list, and if a directory is given, then each file of the directory is put on the list. Therefore, after this step on, there is always a list with at least one file path. Does it makes sense?

@auphofBSF I'll add your suggestions to the docs and update the PR accordingly

**Comment by rola93** [comment]
me too, it's a pleasure to contribute! :D let's wait for @OlafenwaMoses  to merge it

File: imageai/Detection/Custom/__init__.py [hunk 3]
```diff
@@ -507,7 +508,7 @@ def _create_callbacks(self, saved_weights_name, model_to_save):
 
         checkpoint = CustomModelCheckpoint(
             model_to_save=model_to_save,
-            filepath=saved_weights_name + 'ex-{epoch:02d}--loss-{loss:.2f}.h5',
+            filepath=saved_weights_name + 'ex-{epoch:03d}--loss-{loss:8.3f}.h5',
```
Comment (auphofBSF) [inline imageai/Detection/Custom/__init__.py]: Filenames have now had spaces introduced in the loss part 
'detection_model-ex-003--loss-   6.306.h5'
would suggest {loss:08.3f}
otherwise  the changes are functioning OK on the hololens example
TEST OK  with path to models  and a single model file name  
trainer.evaluateModel(model_path="hololens/models"
trainer.evaluateModel(model_path="hololens/models/detection_model-ex-003--loss-   6.306.h5"

**Comment by OlafenwaMoses** [comment]
> > This is a brilliant work. Thanks for the effort on this. It will be better if you can effect the following changes:
> > 
> > * add `print` of  **mAP** evaluation result for each model if a directory path is parsed
> 
> ### Suggestions for updating documentation,
> * @rola93 do you want to incorporate in your PR #302
> * @OlafenwaMoses is the only documentation location ?
> 
> https://github.com/OlafenwaMoses/ImageAI/blob/master/imageai/Detection/Custom/CUSTOMDETECTIONTRAINING.md#evaluating-your-saved-detection-models-map
> 
> ```
> from imageai.Detection.Custom import DetectionModelTrainer
> 
> trainer = DetectionModelTrainer()
> trainer.setModelTypeAsYOLOv3()
> trainer.setDataDirectory(data_directory="/content/hololens")
> evaluate_metrics = trainer.evaluateModel(model_path="/content/hololens/models", 
>                                json_path="/content/hololens/json/detection_config.json", 
>                                iou_threshold=0.5, 
>                                object_threshold=0.3, 
>                                nms_threshold=0.5)
> ```
> 
> ### Console stdout and stderr
> ```
> Using TensorFlow backend.
> Starting Model evaluation....
> WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
> Instructions for updating:
> Colocations handled automatically by placer.
> WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/imageai-2.1.3-py3.6.egg/imageai/Detection/Custom/yolo.py:24: to_float (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
> Instructions for updating:
> Use tf.cast instead.
> /usr/local/lib/python3.6/dist-packages/keras/engine/saving.py:292: UserWarning: No training configuration found in save file: the model was *not* compiled. Compile it manually.
>   warnings.warn('No training configuration found in save file: '
> Model File:  /content/hololens/models/detection_model-ex-02--loss-7.47.h5 
> 
> Using IoU :  0.5
> Using Object Threshold :  0.3
> Using Non-Maximum Suppression :  0.5
> hololens: 0.8252
> mAP: 0.8252
> ===============================
> Model File:  /content/hololens/models/detection_model-ex-01--loss-25.31.h5 
> 
> Using IoU :  0.5
> Using Object Threshold :  0.3
> Using Non-Maximum Suppression :  0.5
> hololens: 0.4390
> mAP: 0.4390
> ===============================
> Model File:  /content/hololens/models/detection_model-ex-03--loss-6.20.h5 
> 
> Using IoU :  0.5
> Using Object Threshold :  0.3
> Using Non-Maximum Suppression :  0.5
> hololens: 0.8415
> mAP: 0.8415
> ===============================
> ```
> 
> ### Dictionary view of metrics
> `evaluate_metrics`
> 
> ```
> [{'average_precision': {'hololens': 0.8252334437735249},
>   'map': 0.8252334437735249,
>   'model_file': '/content/hololens/models/detection_model-ex-02--loss-7.47.h5',
>   'using_iou': 0.5,
>   'using_non_maximum_suppression': 0.5,
>   'using_object_threshold': 0.3},
>  {'average_precision': {'hololens': 0.4389696930319986},
>   'map': 0.4389696930319986,
>   'model_file': '/content/hololens/models/detection_model-ex-01--loss-25.31.h5',
>   'using_iou': 0.5,
>   'using_non_maximum_suppression': 0.5,
>   'using_object_threshold': 0.3},
>  {'average_precision': {'hololens': 0.8414673569376486},
>   'map': 0.8414673569376486,
>   'model_file': '/content/hololens/models/detection_model-ex-03--loss-6.20.h5',
>   'using_iou': 0.5,
>   'using_non_maximum_suppression': 0.5,
>   'using_object_threshold': 0.3}]
> ```
> 
> ### Example of use
> ```
> for metric in evaluate_metrics:
>   print("Model:{modelFile} \t Average mAP:{mAP:.4f}".format(modelFile=metric['model_file'],mAP=metric['map']))
> 
> print('--------------------------------------------------------------------')
> for metric in evaluate_metrics:
>   print("Model:{modelFile}".format(modelFile=metric['model_file']))
>   for theclass in metric['average_precision']:
>       print("\t\t\t\t\t\t\t\t Item:{item} \t mAP:{mAP:.4f}".format(item=theclass, mAP=metric['average_precision'][theclass]))
> ```
> 
> ```
> Model:/content/hololens/models/detection_model-ex-02--loss-7.47.h5 	 Average mAP:0.8252
> Model:/content/hololens/models/detection_model-ex-01--loss-25.31.h5 	 Average mAP:0.4390
> Model:/content/hololens/models/detection_model-ex-03--loss-6.20.h5 	 Average mAP:0.8415
> --------------------------------------------------------------------
> Model:/content/hololens/models/detection_model-ex-02--loss-7.47.h5
> 								 Item:hololens 	 mAP:0.8252
> Model:/content/hololens/models/detection_model-ex-01--loss-25.31.h5
> 								 Item:hololens 	 mAP:0.4390
> Model:/content/hololens/models/detection_model-ex-03--loss-6.20.h5
> 								 Item:hololens 	 mAP:0.8415
> ```

Good work! The update to the README is just fine.

**Comment by OlafenwaMoses** [comment]
> > @auphofBSF, I'm sorry but I didn't get your point with respect to updating the docs
> > I was wondering if shouldn't add a `verbose` parameter to avoid printing always, what do you think?
> 
> I like your thinking but would like to suggest the following to improve overall monitoring
> (@OlafenwaMoses I would appreciate your thoughts on this)
> 
> 1. **A more consistent  and multi-functional global level of verbosity**,  `print(...)`  in project be converted to using the python logging. module
> 
> ```
> import logging
> logging.info(......)
> logging.debug(.....)
> ```
> 
> In that way verbosity can be globally configured for each application usage of imageai and granularly adjusted for each class with a diverse set of logging options up to the final user
> 
> 1. **More detailed progress metrics**, I was looking through all the code for custom object detection and looking for some more granular level of metrics on training or evaluation progress.  I extensively use the module [prometheus_client](https://github.com/prometheus/client_python) `import   prometheus_client`   for monitoring of long running python applications, its lightweight and easy to add and just works.
> 
> I look forward to other suggestion and possibly putting this on feature_request, or enhancements list and to other suggestions and willing to help put some of this in place

@auphofBSF I do agree with your suggestion on using the **logging** feature instead of the print. Will update the library soon to include this.
